tests/save_load.test.js

```diff
@@ -0,0 +1,153 @@
+import { load, save } from '../src/lib/save_load';
+
+test('Save saveAsJSON should create a blob url and download it', async () => {
+  const createObjectURL = jest.fn();
+  const revokeObjectURL = jest.fn();
+  window.URL = { createObjectURL, revokeObjectURL };
+
+  const download = jest.fn();
+  const addListener = jest.fn();
+  const removeListener = jest.fn();
+  window.browser = {
+    downloads: {
+      download,
+      onChanged: { addListener, removeListener },
+    },
+  };
+
+  createObjectURL.mockReturnValue('fake-blob-url');
+
+  const promiseRevokedURL = new Promise((resolve) => {
+    revokeObjectURL.mockImplementation(() => resolve());
+  });
+
+  let listener;
+  addListener.mockImplementation((callback) => {
+    listener = callback;
+  });
+
+  download.mockImplementation(() => {
+    const id = 123;
+    setTimeout(() => {
+      listener({
+        state: { current: 'in_progress' },
+        id,
+      });
+      listener({
+        state: { current: 'complete' },
+        id,
+      });
+    }, 0);
+    return Promise.resolve(id);
+  });
+
+  const promiseRemovedListener = new Promise((resolve) => {
+    removeListener.mockImplementation(() => resolve());
+  });
+
+  const isDownloadComplete = await save.saveAsJSON([{ prop1: 'log1' }]);
+
+  await Promise.all([promiseRevokedURL, promiseRemovedListener]);
+
```

Comment:
since `createObjectURL` has been passed with blob object. I have added a test like the following: 
`expect(createObjectURL).toHaveBeenCalledWith(new Blob());`

Reply:
This doesn't verify the content. I'm surprised that the test would pass, because `expect(createObjectURL).toHaveBeenCalledWith(new Blob());` looks like something that should never pass (`new Blob()` creates a new instance of an opaque object). Does the test really pass if you write it like this?

The test should verify that the file content has been written as expected.

Reply:
yes, test passes with it

Reply:
It also passes if I pass the same content,
`expect(createObjectURL).toHaveBeenCalledWith(
    new Blob([JSON.stringify([{ prop1: 'log1' }])])
  );
`

Reply:
Can you try to figure out why the test unexpectedly passes when an obviously non-matching object is passed? Tests would be rather useless if they claim two inequal things to be equal...

E.g. is there a default matcher that causes any Blob to match any other Blob? (the `expect` API is documented here: https://jestjs.io/docs/en/expect.html ).

Ping me on Matrix (or comment here) if you can't figure it out, then I'll take a look.

Reply:
I just looked up Jest's source code, starting from the definition of `toHaveBeenCalledWith`: https://github.com/facebook/jest/blob/4a59daa8715bde6a1b085ff7f4140f3a337045aa/packages/expect/src/spyMatchers.ts#L1130
It is implemented by a function factory at `createToBeCalledWithMatcher`.

... with intern determines argument equality with `isEqualCall` at https://github.com/facebook/jest/blob/4a59daa8715bde6a1b085ff7f4140f3a337045aa/packages/expect/src/spyMatchers.ts#L575

... and `isEqualCall` ultimately uses `equals` to determine equality: https://github.com/facebook/jest/blob/4a59daa8715bde6a1b085ff7f4140f3a337045aa/packages/expect/src/spyMatchers.ts#L62

The logic is implemented in this file: https://github.com/facebook/jest/blob/4a59daa8715bde6a1b085ff7f4140f3a337045aa/packages/expect/src/jasmineUtils.ts#L30

The relevant part of it is that for unknown objects, that it checks whether the internal class is the same (which is the case for comparing Blob and Blob instances, but not Blob and Object instances), and that every **[enumerable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties)** property has the same value. The `size` and `type` properties of `Blob` instances are not enumerable, so this effectively means that Jest always considers two `Blob` instances equal.